File: Nori/Model/KindDetector.swift
import Foundation

/// Heuristics that decide whether a piece of text is really a link, a color or code.
enum KindDetector {
    private static let linkSchemes: Set<String> = ["http", "https", "ftp", "ftps", "mailto", "sftp", "ssh", "git"]

    /// A single-token URL with a real host (or a mailto: address).
    static func link(in text: String) -> URL? {
        guard !text.isEmpty, text.count <= 2_048, !text.contains(where: \.isWhitespace) else { return nil }
        guard let url = URL(string: text), let scheme = url.scheme?.lowercased(), linkSchemes.contains(scheme) else {
            return nil
        }
        if scheme == "mailto" {
            return text.contains("@") ? url : nil
        }
        guard let host = url.host(), host.contains(".") || host == "localhost" else { return nil }
        return url
    }

    nonisolated(unsafe) private static let hexColor = #/^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{4}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$/#
    nonisolated(unsafe) private static let rgbColor = #/^rgba?\(\s*(\d{1,3})\s*[, ]\s*(\d{1,3})\s*[, ]\s*(\d{1,3})\s*(?:[,/]\s*(0?\.\d+|1(?:\.0+)?|\d{1,3}%)\s*)?\)$/#
    nonisolated(unsafe) private static let hslColor = #/^hsla?\(\s*(\d{1,3}(?:\.\d+)?)\s*[, ]\s*(\d{1,3}(?:\.\d+)?)%\s*[, ]\s*(\d{1,3}(?:\.\d+)?)%\s*(?:[,/]\s*(0?\.\d+|1(?:\.0+)?|\d{1,3}%)\s*)?\)$/#

    /// Normalised `#RRGGBB` or `#RRGGBBAA` when the text is a CSS-style color literal.
    static func color(in text: String) -> String? {
        let value = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard value.count <= 40 else { return nil }

        if (try? hexColor.wholeMatch(in: value)) != nil {
            let digits = String(value.dropFirst())
            switch digits.count {
            case 3, 4:
                return "#" + digits.map { "\($0)\($0)" }.joined().uppercased()
            default:
                return "#" + digits.uppercased()
            }
        }

        if let match = try? rgbColor.wholeMatch(in: value) {
            guard let r = Int(match.1), let g = Int(match.2), let b = Int(match.3), r <= 255, g <= 255, b <= 255 else {
                return nil
            }
            return hexString(r: r, g: g, b: b, alpha: match.4.flatMap { alphaComponent(String($0)) })
        }

        if let match = try? hslColor.wholeMatch(in: value) {
            guard let h = Double(match.1), let s = Double(match.2), let l = Double(match.3), s <= 100, l <= 100 else {
                return nil
            }
            let (r, g, b) = hslToRGB(h: h.truncatingRemainder(dividingBy: 360), s: s / 100, l: l / 100)
            return hexString(r: r, g: g, b: b, alpha: match.4.flatMap { alphaComponent(String($0)) })
        }

        return nil
    }

    private static func alphaComponent(_ raw: String) -> Int? {
        if raw.hasSuffix("%"), let percent = Double(raw.dropLast()) {
            return Int((min(max(percent, 0), 100) / 100 * 255).rounded())
        }
        if let fraction = Double(raw) {
            return Int((min(max(fraction, 0), 1) * 255).rounded())
        }
        return nil
    }

    private static func hexString(r: Int, g: Int, b: Int, alpha: Int?) -> String {
        if let alpha, alpha < 255 {
            return String(format: "#%02X%02X%02X%02X", r, g, b, alpha)
        }
        return String(format: "#%02X%02X%02X", r, g, b)
    }

    private static func hslToRGB(h: Double, s: Double, l: Double) -> (Int, Int, Int) {
        let c = (1 - abs(2 * l - 1)) * s
        let hp = h / 60
        let x = c * (1 - abs(hp.truncatingRemainder(dividingBy: 2) - 1))
        let (r1, g1, b1): (Double, Double, Double) = switch hp {
        case 0..<1: (c, x, 0)
        case 1..<2: (x, c, 0)
        case 2..<3: (0, c, x)
        case 3..<4: (0, x, c)
        case 4..<5: (x, 0, c)
        default: (c, 0, x)
        }
        let m = l - c / 2
        func channel(_ v: Double) -> Int { Int(((v + m) * 255).rounded()) }
        return (channel(r1), channel(g1), channel(b1))
    }

    private static let codeLineStarts: [String] = [
        "$ ", "#!/", "import ", "from ", "func ", "def ", "class ", "struct ", "const ", "let ", "var ", "fn ",
        "package ", "use ", "#include", "SELECT ", "<?xml", "<!DOCTYPE", "enum ", "interface ", "public ", "private ",
        "export ", "@main", "<?php", "using ", "namespace ", "<html", "<div", "<svg",
    ]

    /// Apps whose copies are usually code.
    static let editorBundlePrefixes: [String] = [
        "com.apple.dt.Xcode", "com.microsoft.VSCode", "com.todesktop.230313mzl4w4u92", "dev.zed.Zed",
        "com.jetbrains.", "com.apple.Terminal", "com.googlecode.iterm2", "dev.warp.Warp-Stable",
        "com.mitchellh.ghostty", "com.sublimetext.", "org.vim.", "com.neovide.", "com.github.atom",
    ]

    /// Conservative code detection: needs at least two lines and a score of 3 from independent
    /// signals, so prose with a stray semicolon stays "text".
    static func looksLikeCode(_ text: String, sourceBundleID: String? = nil, isRichText: Bool = false) -> Bool {
        let lines = text.split(omittingEmptySubsequences: false, whereSeparator: \.isNewline).map { String($0) }
        let nonEmpty = lines.filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
        guard nonEmpty.count >= 2 else { return isSingleLineCommand(text) }

        var score = 0

        if let sourceBundleID, editorBundlePrefixes.contains(where: { sourceBundleID.hasPrefix($0) }) { score += 2 }

        let indented = nonEmpty.filter { $0.hasPrefix("  ") || $0.hasPrefix("\t") }.count
        if Double(indented) / Double(nonEmpty.count) >= 0.3 { score += 2 }

        let symbolCount = text.filter { "{}();=<>".contains($0) }.count
        if Double(symbolCount) >= 1.5 * Double(nonEmpty.count) { score += 1 }

        if let first = nonEmpty.first?.trimmingCharacters(in: .whitespaces),
           codeLineStarts.contains(where: { first.hasPrefix($0) }) { score += 1 }

        let punctuationEndings = nonEmpty.filter { line in
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            return trimmed.hasSuffix(";") || trimmed.hasSuffix("{")
        }.count
        if punctuationEndings >= 2 { score += 1 }

        if isRichText { score -= 2 }

        return score >= 3
    }

    private static func isSingleLineCommand(_ text: String) -> Bool {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count <= 300, !trimmed.contains(where: \.isNewline) else { return false }
        let shellStarts = ["$ ", "npm ", "npx ", "brew ", "git ", "curl ", "docker ", "kubectl ", "xcodebuild ", "sudo ",
                           "pip ", "pip3 ", "cargo ", "go ", "swift ", "python ", "python3 ", "node ", "make ", "cd ",
                           "ls ", "rm ", "cp ", "mv ", "chmod ", "ssh ", "scp ", "tar ", "zip ", "open ", "defaults "]
        return shellStarts.contains { trimmed.hasPrefix($0) } && trimmed.contains(where: { "-/.".contains($0) })
    }
}

File: windows/Nori.Core/KindDetector.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Nori.Core;

/// <summary>Heuristics that decide whether a piece of text is really a link, a color or code.</summary>
public static partial class KindDetector
{
    private static readonly HashSet<string> LinkSchemes = new(StringComparer.OrdinalIgnoreCase)
    {
        "http", "https", "ftp", "ftps", "mailto", "sftp", "ssh", "git",
    };

    /// <summary>A single-token URL with a real host (or a mailto: address).</summary>
    public static Uri? Link(string text)
    {
        if (string.IsNullOrEmpty(text) || text.Length > 2048 || text.Any(char.IsWhiteSpace)) return null;
        if (!Uri.TryCreate(text, UriKind.Absolute, out var url)) return null;
        var scheme = url.Scheme;
        if (!LinkSchemes.Contains(scheme)) return null;
        if (scheme.Equals("mailto", StringComparison.OrdinalIgnoreCase))
        {
            return text.Contains('@') ? url : null;
        }
        var host = url.Host;
        if (string.IsNullOrEmpty(host)) return null;
        if (!host.Contains('.') && !host.Equals("localhost", StringComparison.OrdinalIgnoreCase)) return null;
        return url;
    }

    [GeneratedRegex(@"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{4}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$")]
    private static partial Regex HexColor();

    [GeneratedRegex(@"^rgba?\(\s*(\d{1,3})\s*[, ]\s*(\d{1,3})\s*[, ]\s*(\d{1,3})\s*(?:[,/]\s*(0?\.\d+|1(?:\.0+)?|\d{1,3}%)\s*)?\)$", RegexOptions.IgnoreCase)]
    private static partial Regex RgbColor();

    [GeneratedRegex(@"^hsla?\(\s*(\d{1,3}(?:\.\d+)?)\s*[, ]\s*(\d{1,3}(?:\.\d+)?)%\s*[, ]\s*(\d{1,3}(?:\.\d+)?)%\s*(?:[,/]\s*(0?\.\d+|1(?:\.0+)?|\d{1,3}%)\s*)?\)$", RegexOptions.IgnoreCase)]
    private static partial Regex HslColor();

    /// <summary>Normalised <c>#RRGGBB</c> or <c>#RRGGBBAA</c> when the text is a CSS-style color literal.</summary>
    public static string? Color(string text)
    {
        var value = text.Trim();
        if (value.Length > 40) return null;

        if (HexColor().IsMatch(value))
        {
            var digits = value[1..];
            if (digits.Length is 3 or 4)
            {
                return "#" + string.Concat(digits.Select(c => $"{c}{c}")).ToUpperInvariant();
            }
            return "#" + digits.ToUpperInvariant();
        }

        var rgb = RgbColor().Match(value);
        if (rgb.Success)
        {
            var r = int.Parse(rgb.Groups[1].Value, CultureInfo.InvariantCulture);
            var g = int.Parse(rgb.Groups[2].Value, CultureInfo.InvariantCulture);
            var b = int.Parse(rgb.Groups[3].Value, CultureInfo.InvariantCulture);
            if (r > 255 || g > 255 || b > 255) return null;
            return HexString(r, g, b, rgb.Groups[4].Success ? AlphaComponent(rgb.Groups[4].Value) : null);
        }

        var hsl = HslColor().Match(value);
        if (hsl.Success)
        {
            var h = double.Parse(hsl.Groups[1].Value, CultureInfo.InvariantCulture);
            var s = double.Parse(hsl.Groups[2].Value, CultureInfo.InvariantCulture);
            var l = double.Parse(hsl.Groups[3].Value, CultureInfo.InvariantCulture);
            if (s > 100 || l > 100) return null;
            var (r, g, b) = HslToRgb(h % 360, s / 100, l / 100);
            return HexString(r, g, b, hsl.Groups[4].Success ? AlphaComponent(hsl.Groups[4].Value) : null);
        }

        return null;
    }

    private static int? AlphaComponent(string raw)
    {
        if (raw.EndsWith('%') && double.TryParse(raw[..^1], NumberStyles.Float, CultureInfo.InvariantCulture, out var percent))
        {
            return (int)Math.Round(Math.Clamp(percent, 0, 100) / 100 * 255, MidpointRounding.AwayFromZero);
        }
        if (double.TryParse(raw, NumberStyles.Float, CultureInfo.InvariantCulture, out var fraction))
        {
            return (int)Math.Round(Math.Clamp(fraction, 0, 1) * 255, MidpointRounding.AwayFromZero);
        }
        return null;
    }

    private static string HexString(int r, int g, int b, int? alpha)
    {
        if (alpha is { } a && a < 255)
        {
            return $"#{r:X2}{g:X2}{b:X2}{a:X2}";
        }
        return $"#{r:X2}{g:X2}{b:X2}";
    }

    private static (int R, int G, int B) HslToRgb(double h, double s, double l)
    {
        var c = (1 - Math.Abs(2 * l - 1)) * s;
        var hp = h / 60;
        var x = c * (1 - Math.Abs(hp % 2 - 1));
        var (r1, g1, b1) = hp switch
        {
            < 1 => (c, x, 0.0),
            < 2 => (x, c, 0.0),
            < 3 => (0.0, c, x),
            < 4 => (0.0, x, c),
            < 5 => (x, 0.0, c),
            _ => (c, 0.0, x),
        };
        var m = l - c / 2;
        static int Channel(double v, double m) => (int)Math.Round((v + m) * 255, MidpointRounding.AwayFromZero);
        return (Channel(r1, m), Channel(g1, m), Channel(b1, m));
    }

    /// <summary>Converts a normalised hex string back to channels (for rgb()/hsl() captions).</summary>
    public static (int R, int G, int B, int A)? Channels(string hex)
    {
        if (hex.Length is not (7 or 9) || hex[0] != '#') return null;
        if (!int.TryParse(hex.AsSpan(1, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var r)) return null;
        if (!int.TryParse(hex.AsSpan(3, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var g)) return null;
        if (!int.TryParse(hex.AsSpan(5, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var b)) return null;
        var a = 255;
        if (hex.Length == 9 && !int.TryParse(hex.AsSpan(7, 2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out a)) return null;
        return (r, g, b, a);
    }

    /// <summary>"rgb(255, 107, 53)" for the color card caption.</summary>
    public static string RgbText(string hex)
    {
        if (Channels(hex) is not { } c) return string.Empty;
        return c.A < 255
            ? string.Create(CultureInfo.InvariantCulture, $"rgba({c.R}, {c.G}, {c.B}, {c.A / 255.0:0.##})")
            : string.Create(CultureInfo.InvariantCulture, $"rgb({c.R}, {c.G}, {c.B})");
    }

    /// <summary>"hsl(21, 100%, 60%)" for the expanded color preview.</summary>
    public static string HslText(string hex)
    {
        if (Channels(hex) is not { } c) return string.Empty;
        double r = c.R / 255.0, g = c.G / 255.0, b = c.B / 255.0;
        var max = Math.Max(r, Math.Max(g, b));
        var min = Math.Min(r, Math.Min(g, b));
        var l = (max + min) / 2;
        double h = 0, s = 0;
        var d = max - min;
        if (d > 0)
        {
            s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
            if (max == r) h = (g - b) / d + (g < b ? 6 : 0);
            else if (max == g) h = (b - r) / d + 2;
            else h = (r - g) / d + 4;
            h *= 60;
        }
        return string.Create(CultureInfo.InvariantCulture, $"hsl({Math.Round(h)}, {Math.Round(s * 100)}%, {Math.Round(l * 100)}%)");
    }

    private static readonly string[] CodeLineStarts =
    [
        "$ ", "#!/", "import ", "from ", "func ", "def ", "class ", "struct ", "const ", "let ", "var ", "fn ",
        "package ", "use ", "#include", "SELECT ", "<?xml", "<!DOCTYPE", "enum ", "interface ", "public ", "private ",
        "export ", "@main", "<?php", "using ", "namespace ", "<html", "<div", "<svg",
    ];

    /// <summary>Executables (or name prefixes) whose copies are usually code: editors and terminals.</summary>
    public static readonly IReadOnlyList<string> EditorProcessPrefixes =
    [
        "Code.exe", "Code - Insiders.exe", "devenv.exe", "Cursor.exe", "zed.exe", "idea64.exe", "rider64.exe",
        "pycharm64.exe", "webstorm64.exe", "clion64.exe", "goland64.exe", "phpstorm64.exe", "sublime_text.exe",
        "notepad++.exe", "WindowsTerminal.exe", "cmd.exe", "powershell.exe", "pwsh.exe", "mintty.exe",
        "alacritty.exe", "wezterm-gui.exe", "nvim", "gvim.exe", "vim.exe", "neovide.exe", "conhost.exe",
        // macOS bundle identifiers keep the classifier fixtures shared with the Mac test corpus.
        "com.apple.dt.Xcode", "com.microsoft.VSCode", "com.jetbrains.", "com.apple.Terminal",
    ];

    /// <summary>
    /// Conservative code detection: needs at least two lines and a score of 3 from independent
    /// signals, so prose with a stray semicolon stays "text".
    /// </summary>
    public static bool LooksLikeCode(string text, string? sourceApp = null, bool isRichText = false)
    {
        var lines = SplitLines(text);
        var nonEmpty = lines.Where(l => l.Trim(' ', '\t').Length > 0).ToList();
        if (nonEmpty.Count < 2) return IsSingleLineCommand(text);

        var score = 0;

        if (sourceApp is not null && EditorProcessPrefixes.Any(p => sourceApp.StartsWith(p, StringComparison.OrdinalIgnoreCase))) score += 2;

        var indented = nonEmpty.Count(l => l.StartsWith("  ", StringComparison.Ordinal) || l.StartsWith('\t'));
        if ((double)indented / nonEmpty.Count >= 0.3) score += 2;

        var symbolCount = text.Count(c => "{}();=<>".Contains(c));
        if (symbolCount >= 1.5 * nonEmpty.Count) score += 1;

        var first = nonEmpty[0].Trim(' ', '\t');
        if (CodeLineStarts.Any(s => first.StartsWith(s, StringComparison.Ordinal))) score += 1;

        var punctuationEndings = nonEmpty.Count(l =>
        {
            var trimmed = l.Trim(' ', '\t');
            return trimmed.EndsWith(';') || trimmed.EndsWith('{');
        });
        if (punctuationEndings >= 2) score += 1;

        if (isRichText) score -= 2;

        return score >= 3;
    }

    private static readonly string[] ShellStarts =
    [
        "$ ", "npm ", "npx ", "brew ", "git ", "curl ", "docker ", "kubectl ", "xcodebuild ", "sudo ",
        "pip ", "pip3 ", "cargo ", "go ", "swift ", "python ", "python3 ", "node ", "make ", "cd ",
        "ls ", "rm ", "cp ", "mv ", "chmod ", "ssh ", "scp ", "tar ", "zip ", "open ", "defaults ",
        "dotnet ", "winget ", "choco ", "scoop ", "wsl ", "pwsh ", "dir ", "del ", "copy ", "move ", "Get-", "Set-",
    ];

    private static bool IsSingleLineCommand(string text)
    {
        var trimmed = text.Trim();
        if (trimmed.Length > 300 || trimmed.Any(IsNewline)) return false;
        return ShellStarts.Any(s => trimmed.StartsWith(s, StringComparison.Ordinal)) && trimmed.Any(c => "-/.".Contains(c));
    }

    public static bool IsNewline(char c) => c is '\n' or '\r' or '\u2028' or '\u2029' or '\u0085' or '\u000B' or '\u000C';

    /// <summary>Splits on every Unicode line terminator, treating CRLF as one break.</summary>
    public static List<string> SplitLines(string text)
    {
        var result = new List<string>();
        var start = 0;
        for (var i = 0; i < text.Length; i++)
        {
            if (!IsNewline(text[i])) continue;
            result.Add(text[start..i]);
            if (text[i] == '\r' && i + 1 < text.Length && text[i + 1] == '\n') i++;
            start = i + 1;
        }
        result.Add(text[start..]);
        return result;
    }
}
